### PR TITLE
Allow setting ztunnel K8S Pod property `terminationGracePeriodSeconds` via Helm

### PR DESCRIPTION
**Pull request by bleggett**
This allows Helm configuration of 2 preexisting ztunnel config knobs:

1. Allow `terminationGracePeriodSeconds` (standard Kube param, Kube/our current default is 30, this changes it to helm-defined value + 1) to be set for the ztunnel Pod. This is the duration Kube waits after sending a SIGTERM before sending a more agressive/destructive TERM sig.
   
2. Reuses that Helm-defined value to populate ztunnel's existing `TERMINATION_GRACE_PERIOD` env var, which is related, and is the duration ztunnel waits, after getting a SIGTERM, to finish draining its own proxied connections.

(I'm open to better ways to handle this than just adding one second - having 2 separate values seems potentially more confusing/error-prone - the important thing is that one value is always bigger than the other by a reasonable margin)

### CODE REVIEW

**Review by costinm — COMMENTED** [review]
Is 30s the best default ? Why not 30 min or 5 sec ? If we are making it configurable, we should have some guidance on what an admin should set it to. 

I'm generally against defining a shadow API in helm - but this case seems to be working as intended, it sets k8s install options.

**Review by costinm — COMMENTED** [review]
Looks good to me - just please add few more comments on proper value.

**Comment by bleggett** [comment]
Actually, it's a little less surprising if this is inverted, so the raw value is used for Kube and the -1 val is used for ztunnel.

**Comment by costinm** [comment]
On Mon, May 8, 2023 at 12:33 PM Ben Leggett ***@***.***>
wrote:

> Is 30s the best default ? Why not 30 min or 5 sec ? If we are making it
> configurable, we should have some guidance on what an admin should set it
> to.
>
> 30 sec is the K8S default, if unspecified (AFAICT,
> https://cloud.google.com/blog/products/containers-kubernetes/kubernetes-best-practices-terminating-with-grace
> )
>
I'm not sure this PR has the same behavior as the default - wouldn't the
env variable set on ztunnel make it wait for that interval instead of
stopping on SIGTERM ?

I also don't think a lot of users ( including myself ) understand the
impact of the env variable and termination for a daemon set like ztunnel -
will a new ztunnel
be started in parallel for the 30 sec ? Will the pods on the node stop
receiving since ztunnel is removed from the EndpointSlice ? Even for
deployments the
termination is a tricky area.

IMO the safest upgrade for CNI and ztunnel is combined with the kubelet and
node upgrade and using a node cordon/drain. But now that we support
sessions,
a proper upgrade would wait for the max duration of a termination duration
for any pod on the node. Which may be tricky to implement, but as a
workaround
we can have a much larger value.

It's far more complicated than it seems - I don't know how much we can put
in the comments in values.yaml, but we need to at least mention those
issues.

**Comment by costinm** [comment]
My point was that (unless I'm missing something) this change is not a
no-op. Setting the pod option is a no-op, but the env variable was not set
before and changes the behavior.

**Comment by bleggett** [comment]
> My point was that (unless I'm missing something) this change is not a no-op. Setting the pod option is a no-op, but the env variable was not set before and changes the behavior.

Correct, the pod option is (by default) a no-op, the env var is not, as it changes the termination grace period to (`podOption - 1 sec`) - which means by default the ztunnel drain max wait period is 29 and not 5 seconds.

Do you have a better suggestion? 

If we don't intend to expose `TERMINATION_GRACE_PERIOD` env var from ztunnel we should remove it. 
If we intend to expose it we need to offer a mechanism to set it.
If we need to set it, it needs to be smaller than the related pod timeout, and it would be good to ensure that via Helm constraints.

**Comment by bleggett** [comment]
@costinm RE last comment - do you have an alternate/preferred suggestion here?

The ztunnel env var predates this PR, but is non-configurable without this PR.

**Comment by costinm** [comment]
That's another problem - given how tricky it is to install and restart
ztunnel - do we really want setting the drain interval to be an install
time option,
and adjusting it to require restarting ztunnel ? Including potential node
cordon for users who want to be safe, and draining (with the previous value
since the new value takes effect next time) ?

I understand some of the arguments on not creating a new Istiod helm chart
and cleaning up that mess - but they don't apply to ztunnel,
we don't have a legacy of users who would get confused, or CI/CDs that
somehow use helm as API.

My suggestion is to use ztunnel yaml config, as a regular mounted config
map that is controlled by user and not helm - and not repeat
the mistake of having the config of ztunnel as an install-only option with
changes requiring rollout, draining and all that.

This is the perfect example of why install options are a bad API - I can't
imagine a user who wants to adjust the drain duration will want to
do it by doing a drain with the old duration and going trough the entire
ztunnel restart process and risks so next time ztunnel restarts will
have the new drain setting. That's for the env variable - of course for the
pod setting we don't have options, but that is a maximum,
the env variable is really controlling how long ztunnel will wait.

**Comment by costinm** [comment]
@***@***.*** ***@***.***> -  my last comment is far
broader than this PR.

Assuming ztunnel may be embedded in native CNI ( at least for the purpose
of upgrade - i.e. upgrade with the CNI/kubelet/node, as part of platform)
and not under control of regular CI/CD - I think we should use the same
pattern that kube-system has, where config maps like coredns or
calico are exposed and controllable by user. Or even better - define a
proto and CRD for ambient configs ( I suspect many are not
really specific to ztunnel implementation ).

If I remember correctly, one of the goals was to only make new mistakes -
and not repeat old ones...

**Comment by howardjohn** [comment]
That's fair. Although the Pod level setting we have no choice really, but we could just always set that really high and have it handled in-app by ztunnel.

So you are proposing this is a dynamically-updated configmap read by ztunnel, yes?

Do you have a strong preference on that configmap being in or out of Helm?

**Comment by linsun** [comment]
I personally would prefer setting it via helm, i think nearly everyone can understand how to configure this via reading helm values.yaml file, not needing to read another config map doc (plus finding the doc in the first place).

IIUC, @bleggett 's PR enabled us to set the grace period to 30s for env var & pod, and without his PR the env var is 1s which is very short to drain any active connections.

I don't see a better way to do this, so I'd vote LGTM

I'm tagging a few other @istio/wg-environments-maintainers to chime in if you they want to.

**Comment by bleggett** [comment]
> That's another problem - given how tricky it is to install and restart ztunnel - do we really want setting the drain interval to be an install time option, and adjusting it to require restarting ztunnel ? Including potential node cordon for users who want to be safe, and draining (with the previous value since the new value takes effect next time) ? I understand some of the arguments on not creating a new Istiod helm chart and cleaning up that mess - but they don't apply to ztunnel, we don't have a legacy of users who would get confused, or CI/CDs that somehow use helm as API. My suggestion is to use ztunnel yaml config, as a regular mounted config map that is controlled by user and not helm - and not repeat the mistake of having the config of ztunnel as an install-only option with changes requiring rollout, draining and all that. 

Totally fair, that's a much broader existing problem that ztunnel doesn't have a dynamic config mechanism. I agree it probably should have that in some form, and today it has nothing like that outside of xDS.

If we go the CM route, realistically that CM would still be populated/created by Helm initially, mounted into the `ztunnel` pod, and we would add dynamic config reloading to`ztunnel`.

In which case, this PR still directly feeds into that - exposing `terminationGracePeriodSeconds` as a Helm option is still going to be something we would do there, we would just interpolate that into the initial `ztunnel` CM differently, and once `ztunnel` had the ability to detect and reload on CM changes, it could do what it wanted with the updated value.

**Comment by linsun** [comment]
> If we go the CM route, realistically that CM would still be populated/created by Helm initially, mounted into the `ztunnel` pod, and we would add dynamic config reloading to`ztunnel`.

I tend to agree, and I am not convinced if needs to be something dynamic plus the pod config for termination_grace_period  would require restart ztunnel pod anyway?

**Comment by costinm** [comment]
On Fri, May 12, 2023 at 11:17 AM John Howard ***@***.***>
wrote:

> That's fair. Although the Pod level setting we have no choice really, but
> we could just always set that really high and have it handled in-app by
> ztunnel.
>
> So you are proposing this is a dynamically-updated configmap read by
> ztunnel, yes?
>
> Do you have a strong preference on that configmap being in or out of Helm?
>

I have a very strong preference to have it out of helm - and loaded using
normal config map mount, not watcher.
If ztunnel becomes part of an 'upstream' or platform - users will have no
control over installing/reinstalling ztunnel,
but they can modify the config map - like CoreDNS and quite a few other
platform components in kube-system.

It is probably also fine to push it from XDS - like we do for ProxyConfig -
but I think that would be overkill, overcomplex
and not really needed.

**Comment by costinm** [comment]
On Fri, May 12, 2023 at 11:24 AM Lin Sun ***@***.***> wrote:

> I personally would prefer setting it via helm, i think nearly everyone can
> understand how to configure this via reading helm values.yaml file, not
> needing to read another config map doc (plus finding the doc in the first
> place).
>
That assumes ztunnel will remain a user-managed component, and user will
understand how to safely upgrade it, and
that users prefer dangerous and disruptive restarts of the per-node ztunnel
for the smallest fine tunning.

CRDs and config maps are used broadly in K8S - in particular for things
that don't require reinstalling, as changing this particular config is.

Maybe for Istiod the argument is true and some people prefer doing
revision-based upgrades from CI/CD ( with cluster-wide permissions
as Istio upgrades require ) - but regardless of preferences I hope we agree
it is not a best practice from security, operations, config validations,
etc. At least for ztunnel.



> IIUC, @bleggett <https://github.com/bleggett> 's PR enabled us to set the
> grace period to 30s for env var & pod, and without his PR the env var is 1s
> which is very short to drain any active connections.
>
> I don't see a better way to do this, so I'd vote LGTM
>
I mean - having an optional config map that include this is clearly a
better way as it applies on the current restart, not the next one,
and doesn't require restarting ztunnel to specify how long it should wait
on restart. How would you explain to the user - it's a change
in the future ?

**Comment by howardjohn** [comment]
> CRDs and config maps are used broadly in K8S - in particular for things that don't require reinstalling, as changing this particular config is.

I agree with using CM and mounting it. Not sure why it can't be in the helm chart though. 

If its part of the platform then users just don't use helm. But if they are using helm, they should be able to use helm I would think.

**Comment by costinm** [comment]
On Fri, May 12, 2023 at 11:41 AM Ben Leggett ***@***.***>
wrote:

> That's another problem - given how tricky it is to install and restart
> ztunnel - do we really want setting the drain interval to be an install
> time option, and adjusting it to require restarting ztunnel ? Including
> potential node cordon for users who want to be safe, and draining (with the
> previous value since the new value takes effect next time) ? I understand
> some of the arguments on not creating a new Istiod helm chart and cleaning
> up that mess - but they don't apply to ztunnel, we don't have a legacy of
> users who would get confused, or CI/CDs that somehow use helm as API. My
> suggestion is to use ztunnel yaml config, as a regular mounted config map
> that is controlled by user and not helm - and not repeat the mistake of
> having the config of ztunnel as an install-only option with changes
> requiring rollout, draining and all that.
>
> Totally fair, that's a much broader existing problem that ztunnel doesn't
> have a dynamic config mechanism. I agree it probably should have that in
> some form, and today it has nothing like that outside of xDS.
>

It has a yaml config - I think it can periodically read it or watch for
file changes, rust has libraries for this. We don't need to make it super
complex.


> If we go the CM route, realistically that CM would still be
> populated/created by Helm initially, mounted into the ztunnel pod, and we
> would add dynamic config reloading toztunnel.
>
Maybe it can be populated by Helm initially - but I think it is critical to
behave like core-dns or Calico or all other system config maps, in the sense
that once a user edits the config map - it will not be replaced by the next
upgrade.

Configmaps are intended to represent K8S resources that are almost always
user modifiable - mesh config is the only exception I know
where user configs last until the next helm upgrade.



> In which case, this PR still directly feeds into that - exposing
> terminationGracePeriodSeconds as a Helm option is still going to be
> something we would do there, we would just interpolate that into the
> initial ztunnel CM differently, and once ztunnel had the ability to
> detect and reload on CM changes, it could do what it wanted with the
> updated value
>

I have no concerns with the pod option - just the env variable.

I think we can set the pod option to a very high value ( 3 h ) - so users
don't need to worry, it is a max not the actual drain duration.

**Comment by costinm** [comment]
On Fri, May 12, 2023 at 2:23 PM John Howard ***@***.***>
wrote:

> CRDs and config maps are used broadly in K8S - in particular for things
> that don't require reinstalling, as changing this particular config is.
>
> I agree with using CM and mounting it. Not sure why it can't be in the
> helm chart though.
>
> If its part of the platform then users just don't use helm. But if they
> are using helm, they should be able to use helm I would think.
>
Consistency with other k8s platform settings ( calico, core-dns, etc) ?

I don't mind if helm creates it - the fundamental problem is that user
changes to the config map should not be replaced on the next upgrade.
And if we expect ztunnel to become 'platform' - managed along with the
node, kubelet and CNI - we do need to provide a way for
users to configure things independent of the platform admin.

I don't know if platform admins ( GKE, AWS, etc ) use helm or not for
configuring kubelet, CNI and other system configs, but it is clear
that other system components like core-dns allow some user control by
making changes to the configmap.

**Comment by costinm** [comment]
Also the argument that if they use helm - they should be able to configure
everything using values.yaml may make sense
for Istiod ( because that's what we've traditionally done and set as
expectation ), but it is clearly not the case for
all k8s components - many do use CRDs or user-modifiable config maps as
well. Istiod included - many of our
options ( Secrets, initial CA when intermediary CA is used, etc) are still
independent of helm.

**Comment by bleggett** [comment]
> I have no concerns with the pod option - just the env variable.


I get that - but the latter needs to always be smaller than the former - they are implicitly linked.




> I think we can set the pod option to a very high value ( 3 h ) - so users
> don't need to worry, it is a max not the actual drain duration.

That leads to odd behavior if ztunnel hangs on termination, and defeats the purpose of the K8S API - my point is we shouldn't _have_ opinions on what `terminationGracePeriodSeconds` is set to - it's a K8S lifecycle setting and out of our hands.

ztunnel has to respond to whatever that setting is, and make sure it's termination grace is smaller - it can't do anything else.

**Comment by howardjohn** [comment]
to get some real data instead of preferences of developers on our team, I researched the top 10 Helm charts and whether they put configmaps in the chart or have users configure them externally to Helm. I also included Calico and Cilium, since they are similar to ztunnel (but not top 10):


**configmaps in helm:**
kube-prometheus-stack
ingress-nginx
cert-manager
argocd
prometheus
redis
grafana
kubernetes-dashboard
postgresql
Cilium (not top 10, but included as a CNI)
Calico (not top 10, but included as a CNI)

**configmaps not in helm:**
traefik - but there is no configmaps used at all in the software

---

tl;dr I have not found _any_ helm charts that have users configure configmaps to configure the software outside of helm.

**Comment by costinm** [comment]
On Fri, May 12, 2023 at 2:33 PM Ben Leggett ***@***.***>
wrote:

> I have no concerns with the pod option - just the env variable.
>
> I get that - but the latter needs to always be smaller than the former -
> they are implicitly linked.
>
> I think we can set the pod option to a very high value ( 3 h ) - so users
> don't need to worry, it is a max not the actual drain duration.
>
> That leads to odd behavior if ztunnel hangs on termination, and defeats
> the purpose of the K8S API - my point is we shouldn't *have* opinions on
> what terminationGracePeriodSeconds is set to - it's a K8S lifecycle
> setting and out of our hands.
>
> ztunnel has to respond to whatever that setting is, and make sure it's
> termination grace is smaller - it can't do anything else.
>

I don't know if we can use a projection - or if such a corner case needs
much more than documentation.

But it doesn't need to be smaller - if it is larger the k8s pod setting
takes priority and terminates earlier.

**Comment by bleggett** [comment]
> But it doesn't need to be smaller - if it is larger the k8s pod setting
> takes priority and terminates earlier.

Then let's just remove `TERMINATION_GRACE_PERIOD` from ztunnel entirely, and let K8S the process manager.

Just rely only on `terminationGracePeriodSeconds`

**Comment by costinm** [comment]
Ok, how about: ztunnel install ( and maybe istiod ) should NOT depend or
require helm.

That means vendors who manage the components should not be forced to use
helm. Since Istio supports 3 different install methods -
2 of which may internally as implementation use helm, but expose their own
yaml config and one as a CRD - I think
we have a precedent for this in istio.

I believe most of the examples you gave also provide operators - and the
operators have CRDs that control the
options of the component.

If Calico does create and overrides the configmap using helm - we are in
huge problems, because our instructions
for configuring calico include modifying the configmap. Does it mean the
configs are until next time user does a helm
upgrade on calico ?

**Comment by costinm** [comment]
On Fri, May 12, 2023 at 2:38 PM Ben Leggett ***@***.***>
wrote:

> But it doesn't need to be smaller - if it is larger the k8s pod setting
> takes priority and terminates earlier.
>
> Then let's just remove TERMINATION_GRACE_PERIOD from ztunnel entirely,
> and let K8S the process manager.
>
> Just rely only on terminationGracePeriodSeconds
>
I think this may work - ztunnel still gets the signal and can stop
accepting new connections, maybe send GOAWAY
on existing connections - I don't think the env variable does anything
special. If active connections is 0 - it can exit.

**Comment by howardjohn** [comment]
> If Calico does create and overrides the configmap using helm - we are in huge problems, because our instructions for configuring calico include modifying the configmap. Does it mean the configs are until next time user does a helm upgrade on calico ?

I don't think that instruction is a serious doc - its a hack to get it working - not suitable for real world.

**Comment by costinm** [comment]
On Fri, May 12, 2023 at 2:45 PM John Howard ***@***.***>
wrote:

> If Calico does create and overrides the configmap using helm - we are in
> huge problems, because our instructions for configuring calico include
> modifying the configmap. Does it mean the configs are until next time user
> does a helm upgrade on calico ?
>
> I don't think that instruction is a serious doc - its a hack to get it
> working - not suitable for real world.
>
Does it mean Calico is not supported either ? Just Kind, or what do we
really support ?

At least on GKE Calico seems to be installed by addonmanager, and the CRD
we are changing doesn't appear to be created by helm
or addonmanager - I don't know how it was originally created, but as a
proper CRD with "v1" I would expect it reflects a supported
 API surface, not a hack.

**Comment by bleggett** [comment]
> On Fri, May 12, 2023 at 2:38 PM Ben Leggett ***@***.***> wrote: But it doesn't need to be smaller - if it is larger the k8s pod setting takes priority and terminates earlier. Then let's just remove TERMINATION_GRACE_PERIOD from ztunnel entirely, and let K8S the process manager. Just rely only on terminationGracePeriodSeconds
> I think this may work - ztunnel still gets the signal and can stop accepting new connections, maybe send GOAWAY on existing connections - I don't think the env variable does anything special. If active connections is 0 - it can exit.


Yep. `ztunnel` pod will get `SIGTERM` first from K8S, it can try to drain connections for $UNKNOWN_AMOUNT_OF_TIME (e.g. whatever K8S waits as directed by `terminationGracePeriodSeconds`, and if it doesn't clean everything up by the time K8S checks in again and `SIGKILL`'s or whatever, then too bad.

Totally OK removing the ZT flag and relying on normal platform lifecycle management, I just don't want to keep both and only configure one.

**Comment by costinm** [comment]
Sgtm.

That resolves this PR ( k8s pod setting clearly belong in helm ), but we still need separately to discuss if we mandate helm as the only API and config surface, and how the UX will look like given the risks and costs of reinstalling for simple config changes.

**Comment by bleggett** [comment]
> Sgtm.
> 
> That resolves this PR ( k8s pod setting clearly belong in helm ), but we still need separately to discuss if we mandate helm as the only API and config surface, and how the UX will look like given the risks and costs of reinstalling for simple config changes.

1. Agree, sounds good.

2. I'm moving this discuss over ztunnel's `TERMINATION_GRACE_PERIOD`  to a `ztunnel` repo issue, but one thing that might be a sticking point for removing `TERMINATION_GRACE_PERIOD` from ztunnel proper is it hampers the "outside of K8S" story for ztunnel. 

- If we go with the K8S setting and remove the ztunnel setting ("outside-in" - then ztunnel running outside of K8S would need an external watchdog to do `SIGTERM-wait-SIGKILL`, which seems bad.

**Comment by costinm** [comment]
Or leave TERMINATION_GRACE_PERIOD for 'outside of k8s' - but I'm not
entirely sure I understand how termination works on a VM.

My understanding was that systemd ( or docker if ztunnel runs in a
container ) will also send SIGTERM first. And eventually SIGKILL.
Any draining or termination needs to be initiated by the VM/docker somehow.

If we ( or user ) automates ztunnel upgrade - they will need some tool to
send SIGTERM and SIGKILL, how else can they drain ?

**Comment by bleggett** [comment]
> If we ( or user ) automates ztunnel upgrade - they will need some tool to send SIGTERM and SIGKILL, how else can they drain ?

I'm fine with making that an explicit requirement, and yeah totally that's what e.g. `systemd` does - just making sure everyone is OK with that.

**Comment by bleggett** [comment]
> Or leave TERMINATION_GRACE_PERIOD for 'outside of k8s' - but I'm not entirely sure I understand how termination works on a VM. My understanding was that systemd ( or docker if ztunnel runs in a container ) will also send SIGTERM first. And eventually SIGKILL. Any draining or termination needs to be initiated by the VM/docker somehow. If we ( or user ) automates ztunnel upgrade - they will need some tool to send SIGTERM and SIGKILL, how else can they drain ?

Aight so as per the discuss in https://github.com/istio/ztunnel/issues/519 I think we're at the following point.

1. We can't remove either setting (k8s or ztunnel)- we need both, so ztunnel can have guarantees about the window it has to do graceful RST.
2. We need to set the `ztunnel` property to LESS-THAN the k8s setting, to support 1)
3. That means what's in this PR is still needed:
    - We need to set the k8s property
    - Helm must be used to set the k8s property
    - We need to set the ztunnel property to LESS-THAN the k8s property
4. @costinm brings up the very good point that the ztunnel value in question here should (along with others) _not_ simply be an install-time-only env variable, as they are today.


@costinm are you ok with approving this given the above, and I will create a separate `istio/ztunnel` issue to track the need for dynamic config?

**Comment by costinm** [comment]
I commented on the ztunnel bug as well.

I think we're again moving into NIH territory - inventing a new API ( or 2
when you implement the dynamic one ) without
clear requirements and design - and adding a pretty big operating burden on
our users who will need to understand how
this is supposed to be used.

What is the expected stop process for ztunnel and why ? AFAIK, the only
reasons ztunnel would be stopped is:
- users want to stop using ambient - but this process starts with removing
the annotation on namespaces, restarting
pods or whatever other processes get them to stop using istio. In this case
- it doesn't really matter how ztunnel dies.
- upgrade of ztunnel. I assume that's the case we want to optimize for, to
avoid data loss.

In the upgrade case, so far the options I know are cordoning the node - or
using the CNI layer to move all new connections
to the new upgraded ztunnel instance. For a VM - similar mechanism exist (
start the new process or container first, use
low level networking to redirect new stuff to the new process).

The only 'actionable' thing the old ztunnel can do in response to SIGTERM
is to send GOAWAY and stop accepting -
both will accelerate the move of connections to the new ztunnel. For any
in-progress TCP connection, including long
lived streams - there is nothing ztunnel can do using the env variable.
SIGKILL will send the RST and terminate the
connection non-gracefully, and I don't think there is any other graceful
method ( the GOAWAY on the H2 connection
may signal some smart clients to start gracefully terminating long-lived
connections ).

So going back to this - what do we tell operators to do with this helm (
and future dynamic ) API ? Are we committed to
support this as an API long term - and test and provide this ( TBD )
behavior as part of the ambient contract ? If
ambient gets implemented 'natively' in CNI or network layers - would we
expect them to also support such an API ?

**Comment by costinm** [comment]
BTW - if you really have a need for this PR, happy to approve it, I don't
want to block you - but I don't think the
helm install of ztunnel is part of the 'contract' or 'standard' ambient API
surface ( just like I don't think all helm
options for istio are anything but a helm install API ). I am not trying to
block this PR - the bar for helm
settings and env variables is pretty low and I don't remember many cases
where such changes were blocked.

I do not think we should make it a dynamic API - my point ( which I
probably failed to explain ) was that _if_ we
have a clear use case and expect operators to need to control it, it would
be best if it is a proper CRD or
configmap that operators can actually use independent of how they installed
ztunnel ( or whatever implements
the ambient protocols ).

**Comment by costinm** [comment]
I missed on case for ztunnel termination - when operating as a sidecar, the
termination process needs to be different
and AFAIK it does need the setting ( which is already defined for sidecars
):
- on SIGTERM - continue to accept requests ( persistent sessions, app may
do its own graceful shutdown)
- after the drain duration - which for sidecars can be set per pod - start
GOAWAY and rejecting new connections.
- if active connections get to zero - exit.
- SIGKILL terminates both app and sends RST on remaining connections
( not sure if I got this right - and the quick search didn't find a good
page with how we shutdown envoy, we should probably
add more tests and docs around this )

**Comment by bleggett** [comment]
> BTW - if you really have a need for this PR, happy to approve it, I don't want to block you - but I don't think the helm install of ztunnel is part of the 'contract' or 'standard' ambient API surface ( just like I don't think all helm options for istio are anything but a helm install API ). 

Yeah, I don't think we can legitimately avoid letting people set the K8S prop exposed in this PR. And if we set the K8S prop, under the hood the ztunnel prop needs to be LESS-THAN the k8s prop. 

So we do need the outward-facing change here. The internal configuration of the ztunnel prop we can redo or drop later, but for now it does need to track the k8s value.

Before this PR you couldn't set the `ztunnel` config prop directly via Helm. After this PR you can't set the ztunnel config prop directly via Helm.

Since the only outward-facing thing we are explicitly allowing people to set/expose directly is the K8S property here, I don't think we're painting ourselves into a corner here.

**Comment by bleggett** [comment]
Aight - @costinm @howardjohn I've updated this PR to just set the K8S prop, and created https://github.com/istio/ztunnel/pull/522 to remove the otherwise-dependent ztunnel setting.

**Comment by bleggett** [comment]
@costinm @Monkeyanator @jacob-delgado 

Are we good to approve/merge this?

This is now a very simple PR that simply exposes a standard k8s Pod API prop to the chart.

**Review by howardjohn — APPROVED** [review]

**Review by jacob-delgado — APPROVED** [review]
